### PR TITLE
Forbid private library functions be attached with using for outside their declaration scope

### LINKED ISSUE #13764: Private library functions can be attached and used outside of the library
## Description
Private library functions are normally not accessible outside of the library. `L.f` is actually an invalid expression, both inside and outside of the library, when `f` is private. It appears, however, that it's possible to bypass this restriction by binding a private function to a type at file-level.

This is a bug. Private library functions should not be usable in `using for` outside of the library where they are defined.

## Steps to Reproduce
```solidity
library L {
    function privateFunction(uint) private {}
}

using {L.privateFunction} for uint;

contract C {
    function test() public {
        uint x;
        x.privateFunction();    // No errors
        //L.privateFunction(x); // Error: Member "privateFunction" not found or not visible after argument-dependent lookup in type(library L).
    }
}
```

## Environment

- Compiler version: 0.8.17 (`develop`)

### PR DESCRIPTION
**Pull request by matheusaaguiar**
fixes #13764 .

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ Bugfixes:
  * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
  * SMTChecker: Fix internal error when using user defined types as mapping indices or struct members.
  * SMTChecker: Fix internal error when deleting struct member of function type.
+ * TypeChecker: Fix bug where private library functions could be attached with  ``using for`` outside their declaration scope.
```
Comment (nikola-matic) [inline Changelog.md]:
```suggestion
 * TypeChecker: Fix bug where private library functions could be attached with  ``using for`` outside of their declaration scope.
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -3816,6 +3816,20 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
...
+		if (
+			functionDefinition.visibility() == Visibility::Private &&
+			functionDefinition.scope() != m_currentContract
+		)
+			m_errorReporter.typeError(
+				6772_error,
+				path->location(),
+				fmt::format(
+					"Function \"{}\" is private and therefore cannot be attached"
+					" to a type outside the library where it is defined.",
```
Comment (nikola-matic) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
					" to a type outside of the library where it is defined.",
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -3816,6 +3816,20 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 				)
 			);
 
+		if (
+			functionDefinition.visibility() == Visibility::Private &&
+			functionDefinition.scope() != m_currentContract
+		)
+			m_errorReporter.typeError(
+				6772_error,
+				path->location(),
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:3831]: Would be nice to have a secondary location pointing at the function declaration.

```suggestion
				path->location(),
				SecondarySourceLocation().append(
						"Function defined here:",
						functionDefinition.location()
				),
```

Also for the other error above.

And make sure to check the output by manually compiling something on the CLI. The secondary messages do not show up in isoltest output.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -3816,6 +3816,20 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
...
+				fmt::format(
+					"Function \"{}\" is private and therefore cannot be attached"
+					" to a type outside of the library where it is defined.",
+					joinHumanReadable(path->path(), ".")
+				)
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:3840]: I'd add an assert inside the if that the function is a library function. If we ever allow non-contract functions to be private, the message will need to be reworded.

File: Changelog.md [hunk 5]
```diff
@@ -30,6 +30,7 @@ Bugfixes:
  * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
  * SMTChecker: Fix internal error when using user defined types as mapping indices or struct members.
  * SMTChecker: Fix internal error when deleting struct member of function type.
+ * TypeChecker: Fix bug where private library functions could be attached with  ``using for`` outside of their declaration scope.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
```

**Review by cameel — DISMISSED** [review]
Looks good so approving but there's some minor stuff that would be nice to have.

**Comment by matheusaaguiar** [comment]
Addressed @cameel suggestions and rebased on develop.

**Review by cameel — APPROVED** [review]

**Pull request merged**